teps/0031-tekton-bundles-cli.md

```diff
@@ -66,3 +71,95 @@ Provide a way for users to...
+
+```shell
+# REF = Docker-like Tekton Bundle reference, eg. docker.io/myworkspace/mybundle:1.0
+# BUNDLE_OBJECT = path to a file or directory or STDIN of a valid Tekton Task or Pipeline object to be included in the
+# bundle.
+
+# Generates a Tekton Bundle from a set of 1+ bundle objects. Like `docker`, tags the object with the specified reference
+# locally.
+tkn remote build <REF> [BUNDLE_OBJECT...]
+
+# Like docker, publishing the locally referenced Tekton Bundle to a remote repository.
+tkn remote push <REF>
+
+# Fetches a Tekton Bundle and prints its contents in a configurable format. If KIND is specified, will print only
+# objects of the specified kind. If KIND and NAME are specified, will retrieve a specific object.
+tkn remote get <REF> [KIND] [NAME] --output=[FORMAT]
+
+# The following are amendments to existing commands.
+
+# Retrieves from an image the requested task(s) rather than from the cluster.
+tkn task list --image=<REF>
+tkn task get --image=<REF> [NAME]
+
+# Adds the specified BUNDLE_OBJECT to the referenced Tekton Bundle.
+tkn task create --image=<REF> [BUNDLE_OBJECT]
+
+# Removes a named Task from the referenced Tekton Bundle.
+tkn task delete --image=<REF> [NAME]
+
+# Starts a new TaskRun using the Task referenced by this image.
+tkn task start --image=<REF> [NAME]
+
+# ... the pipeline command receives these changes as well
+```
+
+### Detailed Design
+
+To offer a developer experience that users are familiar with, we structure the commands similar to the process of
+building and publishing a Docker image. Namely, we introduce the concept of a "local" Tekton Bundle that is "tagged".
+This enables the user to continue manipulating (adding/removing Tekton Tasks and Pipelines) a Tekton Bundle before
+publishing. This differs from a "remote" Tekton Bundle which has been published to a remote repository like Docker Hub.
+
+This requires the addition of caching to "tkn" to store a certain amount of the Tekton Bundles locally so the user
+doesn't have to fetch them every time. This can be done in a variety of ways (eg printing out the full OCI spec to 
+files).
+
+To accomodate some of the authentication requirements of pushing to a remote repository, we will add global
+configuration options and sensible defaults (like using the common Docker `config.json`) to authenticate into image
+repositories, including private ones. At the top level we will add the following flags (which will only be used when
+Tekton Bundles are the target):
+
+```shell
+# Skips https checks for interacting with remote repositories, ONLY.
+tkn --insecure-skip-verify
+
+# Adds the cert and key to the TLS config used to establish a TLS connection to the repository.
+tkn --repository-client-cert --repository-client-key
+
+# Flags to override the default authentication mechanism which is to use the user's docker config in the default
+# location.
+tkn --repository-auth "bearer=<token">
+# ... or ...
+tkn --repository-auth "basic=<username>:<password"
+```
```

Comment:
Some of those flags may be confusing — `--insecure-skip-verify` for example, does it apply to the connection between `tkn` and the cluster, or something else ?

Reply:
That's a fair point. Should we prefix it with `--bundle` or something like that? Might make it even more verbose. I am open to ideas here (we could also have a .tknrc I suppose).

Reply:
I think like `--insecure` would be a better name, but it should belong to the subcommand, not `tkn` root, so that the extra context is given by the subcommand, e.g.
```
# Insecure refers to the connection to the registry
tkn bundle push --insecure <REF>
```

The problem with that approach is for commands that talk with both the cluster and the registry, e.g.
```
tkn task start --image=<REF> [NAME]
```

The solution could be, when used out of context,  to have a more descriptive argument name:
```
tkn task start --image=<REF> --bundle-insecure  [NAME]  
```

So we could have the following to have insecure communication to the cluster and to the registry:
```
tkn task start --insecure --bundle-insecure --image=<REF>  [NAME]  
```